### PR TITLE
An Introduction to the Crystal Language

### PR DESCRIPTION
**Pull request by rbnpercy**
The Highs & Lows of Crystal Lang - an introduction from a polyglot Dev;  Tech Article.

### CODE REVIEW

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+  mail: "robin@percy.pw"
+  avatar: "https://secure.gravatar.com/avatar/685342d5e7f42c3ab8d251d7d4a53308?s=100&d=mm&r=g"
+tags:
+- Crystal Language
+- Low-Level Programming
+- Systems Programming
+---
+
+
+The Highs and Lows of Crystal  (An Introduction to the Language)
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: This is not needed.

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+====================================
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: Please remove the underlining too.

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+Crystal is a statically-typed, compiled, systems programming language with the aims of being as fast as c/c++, while having a syntax as readable as Ruby.  This article is an introduction to the Crystal Language, through the eyes of a polyglot programmer.  Being a former programmer of both C and Ruby, I have been able to explore the ins-and-outs of Crystal with an objective mindset and give an unbiased opinion on its features; from it's low-level primitives to its familiar syntax, and much inbetween.
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: inbetween => in-between

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+~~~
+
+The fact that you can use Ruby syntax highlighting natively for Crystal says everything!  Coming from Ruby to Crystal is a remarkably easy adaptation.  The fact is; one copy and paste code from Ruby to Crystal and 90% of the time it will run with no errors.  The creators of Crystal understand that Ruby is undoubtedly the most visually appealing language, and therefore built Crystal to take as much influence as possible from a design perspective.  You can even run Crystal programs using the `Ruby` shell command and vice versa since the syntax is valid for both languages!
+
+
+## Binding C
+
+One of the big selling points for Crystal is the ease with which you can interface with C libraries.  "Crystal allows you to bind to existing C libraries without writing a single line in C.  Additionally, it provides some conveniences like `out` and `to_unsafe` so writing bindings is as painless as possible."
+
+Let's build a simple script in C that says "hi!".  We'll then write a Crystal app to bind to our C library.  This is a great start point for anyone who wants to know about binding C in Crystal.
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: start => starting

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+First off, let's create a project with Crystal's scaffolding tool (I'll cover this feature later).  Run:
+
+	$ crystal init app sayhi_c
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: Please enclose the command using the bash syntax like so:


```bash
$ crystal init app sayhi_c
```

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+~~~ yml
+dependencies:
+  github:
+    github: felipeelias/crystal-github
+    version: ~> 0.1.0
+~~~
+
+#### Documentation & Formatting
+
+Crystal has a great inbuilt tool for generating documentation and formatting files.  The documentation that is generated is excellent - built in html/css and almost instantly ready to deploy.
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: inbuilt => built-in

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+To generate documentation, from the project root directory we can simply run:
+
+	$ crystal doc
+
+This will create a docs directory, with a doc/index.html entry point. All files inside the root src directory of the project from which we ran the command will be considered.
+
+Alongside this, the inbuilt *Formatter* tool is a great feature of the language.  We can run the formatter over our project by running:
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: inbuilt => built-in

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+
+	$ crystal tool format
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: Please enclose this command using the bash syntax highlighter.

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+We can use this tool to unify code styles and to submit documentation improvements to Crystal itself.  The formatter is also very fast, so very little time is lost if you format the entire project's codebase instead of just a single file.
+
+*Both of these features are very cool, and highly useful!*
+
+## What I don't like
+
+Like anything in this world, Crystal can't possibly be perfect!  There are two ***very*** small issues that I find with it...
+
+As a Polyglot programmer, I've had to learn a bunch of different programming paradigms.  While this isn't a fault in Crystal, the fact that it's Object Orientated is pretty much the only thing I'm not too keen on in Crystal.  Other than that; being a relatively young &amp; immature language, there's often a lack of documentation available when you want to do something incredibly specific.
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: Object Orientated => object oriented

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+**Seeing this as an opportunity** instead of a foible - it's actually kind of cool, because this means we can write documentation ourselves and hack sample apps together to become early adopters and decent contributors in the Crystal community!
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: Any reason why this is boldly highlighted?

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+**I really rather like it!**
+
+Although relatively immature, Crystal is a promising language with a growing Dev community surrounding it.  In my previous article about Auth0 Lock / Iris Image Recognition, I mentioned the fact that it'd be better to use the pHash / Blockhash libraries for a production environment.  If I was to build that system, I would most definitely use Crystal to bind to those C libraries.  I know that I'd be getting fantastically close-to-C speeds, and with the ease and joy of writing Crystal!
+
+I am very much looking forward to seeing where this language goes.  I think the adoption rate will rapidly increase and I'm excited to see startups using it in production systems.  I am currently experimenting in building a Crystal library for the Auth0 API.  I will write another article on building an API client in Crystal when I'm finished.
+
+I do hope this article has inspired you to give Crystal a try, and look forward to hearing your feedback if/when you do!  If you need any help and want to ask any questions, [reach out to me](mailto:robin@percy.pw) - I'm happy to help!
+
+  ***[ - @rbin](http://twitter.com/rbin)***
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: Better written as =>

If you need any help and want to ask questions, [reach out to me via email][mailto:robin@percy.pw) and [@rbin](https://twitter.com/rbin) on twitter, I'm happy to help!

Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: So you can eliminate the twitter handle you put on a separate line.

File: _posts/2017-06-21-an-introduction-to-crystal-lang.markdown [hunk 12]
```diff
@@ -0,0 +1,316 @@
+---
+layout: post
+title: "The Highs & Lows of Crystal - an intro to Crystal Lang"
```
Comment (unicodeveloper) [inline _posts/2017-06-21-an-introduction-to-crystal-lang.markdown]: => "The Highs & Lows of Crystal - An Introduction to Crystal Language"

**Comment by rbnpercy** [comment]
I have committed the article with all comments addressed.  Really appreciate the feedback! :+1:

**Pull request merged**